File: harness.js
const express = require('express');
const fs = require('fs');
const path = require('path');
const { execFileSync } = require('child_process');

const app = express();
app.use(express.json({ limit: '2mb' }));
app.use(express.urlencoded({ extended: true }));

function logWarn(msg) {
  console.warn(`[harness] ${msg}`);
}

function findDotnetExe() {
  try {
    execFileSync('dotnet', ['--info'], { stdio: 'ignore' });
    return 'dotnet';
  } catch (_) {
    return null;
  }
}

function compileHarness() {
  const dotnet = findDotnetExe();
  if (!dotnet) throw new Error('dotnet not available');

  const tempDir = path.join('/tmp', `ordering-harness-${Date.now()}`);
  fs.mkdirSync(tempDir, { recursive: true });

  const csproj = `
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <OutputType>Exe</OutputType>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>
</Project>`.trim();

  const program = `
using System.Text.Json;
using System.Text.Json.Serialization;

var mode = args.Length > 0 ? args[0] : "";
var json = Console.In.ReadToEnd();
var opts = new JsonSerializerOptions
{
    PropertyNameCaseInsensitive = true,
    DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
};

try
{
    switch (mode)
    {
        case "createorderdraftcommandhandler-handle":
        {
            var message = DeserializeCreateOrderDraftCommandInput(json, opts) ?? new CreateOrderDraftCommandInput();
            var items = message.Items ?? Array.Empty<BasketItem>();

            var handler = new CreateOrderDraftCommandHandler();
            var cmd = new CreateOrderDraftCommand(message.BuyerId ?? "", items);
            var result = await handler.Handle(cmd, CancellationToken.None);

            Console.Write(JsonSerializer.Serialize(result, opts));
            break;
        }

        case "basketitemextensions-toorderitemdto":
        {
            var item = DeserializeBasketItemInput(json, opts) ?? new BasketItem();
            var result = BasketItemExtensions.ToOrderItemDTO(item);
            Console.Write(JsonSerializer.Serialize(result, opts));
            break;
        }

        case "basketitemextensions-toorderitemsdto":
        {
            var items = DeserializeBasketItemsInput(json, opts);
            var result = BasketItemExtensions.ToOrderItemsDTO(items).ToArray();
            Console.Write(JsonSerializer.Serialize(result, opts));
            break;
        }

        default:
            throw new Exception("Unknown mode");
    }
}
catch (Exception ex)
{
    Console.Error.WriteLine(ex.ToString());
    Environment.Exit(1);
}

static CreateOrderDraftCommandInput? DeserializeCreateOrderDraftCommandInput(string json, JsonSerializerOptions opts)
{
    if (string.IsNullOrWhiteSpace(json) || json.Trim() == "null")
        return null;

    using var doc = JsonDocument.Parse(json);
    var root = doc.RootElement;

    if (root.ValueKind != JsonValueKind.Object)
        throw new JsonException("Expected object input for CreateOrderDraftCommandHandler.Handle.");

    if (TryGetPropertyIgnoreCase(root, "message", out var messageProp) && messageProp.ValueKind == JsonValueKind.Object)
        return JsonSerializer.Deserialize<CreateOrderDraftCommandInput>(messageProp.GetRawText(), opts);

    return JsonSerializer.Deserialize<CreateOrderDraftCommandInput>(json, opts);
}

static BasketItem? DeserializeBasketItemInput(string json, JsonSerializerOptions opts)
{
    if (string.IsNullOrWhiteSpace(json) || json.Trim() == "null")
        return null;

    using var doc = JsonDocument.Parse(json);
    var root = doc.RootElement;

    if (root.ValueKind == JsonValueKind.Object)
    {
        if (TryGetPropertyIgnoreCase(root, "item", out var itemProp) && itemProp.ValueKind == JsonValueKind.Object)
            return JsonSerializer.Deserialize<BasketItem>(itemProp.GetRawText(), opts);

        return JsonSerializer.Deserialize<BasketItem>(json, opts);
    }

    throw new JsonException("Expected BasketItem object input.");
}

static BasketItem[] DeserializeBasketItemsInput(string json, JsonSerializerOptions opts)
{
    if (string.IsNullOrWhiteSpace(json) || json.Trim() == "null")
        return Array.Empty<BasketItem>();

    using var doc = JsonDocument.Parse(json);
    var root = doc.RootElement;

    if (root.ValueKind == JsonValueKind.Array)
    {
        return JsonSerializer.Deserialize<BasketItem[]>(json, opts) ?? Array.Empty<BasketItem>();
    }

    if (root.ValueKind == JsonValueKind.Object)
    {
        if (TryGetPropertyIgnoreCase(root, "basketItems", out var basketItems) && basketItems.ValueKind == JsonValueKind.Array)
            return JsonSerializer.Deserialize<BasketItem[]>(basketItems.GetRawText(), opts) ?? Array.Empty<BasketItem>();

        if (TryGetPropertyIgnoreCase(root, "items", out var items) && items.ValueKind == JsonValueKind.Array)
            return JsonSerializer.Deserialize<BasketItem[]>(items.GetRawText(), opts) ?? Array.Empty<BasketItem>();
    }

    throw new JsonException("Expected BasketItem[] input or wrapper object containing basketItems/items.");
}

static bool TryGetPropertyIgnoreCase(JsonElement element, string name, out JsonElement value)
{
    foreach (var prop in element.EnumerateObject())
    {
        if (string.Equals(prop.Name, name, StringComparison.OrdinalIgnoreCase))
        {
            value = prop.Value;
            return true;
        }
    }

    value = default;
    return false;
}

public sealed class CreateOrderDraftCommandHandler
{
    public Task<OrderDraftDTO> Handle(CreateOrderDraftCommand message, CancellationToken cancellationToken)
    {
        var order = Order.NewDraft();
        var orderItems = message.Items.Select(i => i.ToOrderItemDTO());
        foreach (var item in orderItems)
        {
            order.AddOrderItem(item.ProductId, item.ProductName, item.UnitPrice, item.Discount, item.PictureUrl, item.Units);
        }

        return Task.FromResult(OrderDraftDTO.FromOrder(order));
    }
}

public static class BasketItemExtensions
{
    public static IEnumerable<OrderItemDTO> ToOrderItemsDTO(this IEnumerable<BasketItem> basketItems)
    {
        foreach (var item in basketItems)
        {
            yield return item.ToOrderItemDTO();
        }
    }

    public static OrderItemDTO ToOrderItemDTO(this BasketItem item)
    {
        return new OrderItemDTO()
        {
            ProductId = item.ProductId,
            ProductName = item.ProductName,
            PictureUrl = item.PictureUrl,
            UnitPrice = item.UnitPrice,
            Units = item.Quantity
        };
    }
}

public sealed record CreateOrderDraftCommand(string BuyerId, IEnumerable<BasketItem> Items);

public sealed class BasketItem
{
    public string? Id { get; init; }
    public int ProductId { get; init; }
    public string? ProductName { get; init; }
    public decimal UnitPrice { get; init; }
    public decimal OldUnitPrice { get; init; }
    public int Quantity { get; init; }
    public string? PictureUrl { get; init; }
}

public sealed record OrderDraftDTO
{
    public IEnumerable<OrderItemDTO> OrderItems { get; init; } = Array.Empty<OrderItemDTO>();
    public decimal Total { get; init; }

    public static OrderDraftDTO FromOrder(Order order)
    {
        return new OrderDraftDTO()
        {
            OrderItems = order.OrderItems.Select(oi => new OrderItemDTO
            {
                Discount = oi.Discount,
                ProductId = oi.ProductId,
                UnitPrice = oi.UnitPrice,
                PictureUrl = oi.PictureUrl,
                Units = oi.Units,
                ProductName = oi.ProductName
            }),
            Total = order.GetTotal()
        };
    }
}

public sealed record OrderItemDTO
{
    public int ProductId { get; init; }
    public string? ProductName { get; init; }
    public decimal UnitPrice { get; init; }
    public decimal Discount { get; init; }
    public int Units { get; init; }
    public string? PictureUrl { get; init; }
}

public sealed class Order
{
    private readonly List<OrderItem> _orderItems = new();
    private bool _isDraft;

    public IReadOnlyCollection<OrderItem> OrderItems => _orderItems.AsReadOnly();

    public static Order NewDraft()
    {
        var order = new Order
        {
            _isDraft = true
        };
        return order;
    }

    public void AddOrderItem(int productId, string? productName, decimal unitPrice, decimal discount, string? pictureUrl, int units = 1)
    {
        var existingOrderForProduct = _orderItems.SingleOrDefault(o => o.ProductId == productId);

        if (existingOrderForProduct != null)
        {
            if (discount > existingOrderForProduct.Discount)
            {
                existingOrderForProduct.SetNewDiscount(discount);
            }

            existingOrderForProduct.AddUnits(units);
        }
        else
        {
            var orderItem = new OrderItem(productId, productName, unitPrice, discount, pictureUrl, units);
            _orderItems.Add(orderItem);
        }
    }

    public decimal GetTotal() => _orderItems.Sum(o => o.Units * o.UnitPrice);
}

public sealed class OrderItem
{
    public int ProductId { get; }
    public string? ProductName { get; }
    public decimal UnitPrice { get; }
    public decimal Discount { get; private set; }
    public int Units { get; private set; }
    public string? PictureUrl { get; }

    public OrderItem(int productId, string? productName, decimal unitPrice, decimal discount, string? pictureUrl, int units)
    {
        if (units <= 0)
        {
            throw new Exception("Invalid number of units");
        }

        if ((unitPrice * units) < discount)
        {
            throw new Exception("The total of order item is lower than applied discount");
        }

        ProductId = productId;
        ProductName = productName;
        UnitPrice = unitPrice;
        Discount = discount;
        PictureUrl = pictureUrl;
        Units = units;
    }

    public void SetNewDiscount(decimal discount)
    {
        if (discount < 0)
        {
            throw new Exception("Discount is not valid");
        }

        Discount = discount;
    }

    public void AddUnits(int units)
    {
        if (units < 0)
        {
            throw new Exception("Invalid units");
        }

        Units += units;
    }
}

public sealed class CreateOrderDraftCommandInput
{
    public string? BuyerId { get; init; }
    public BasketItem[]? Items { get; init; }
}
`.trim();

  fs.writeFileSync(path.join(tempDir, 'Harness.csproj'), csproj);
  fs.writeFileSync(path.join(tempDir, 'Program.cs'), program);

  try {
    execFileSync(
      dotnet,
      ['build', path.join(tempDir, 'Harness.csproj'), '-c', 'Release', '-o', path.join(tempDir, 'out')],
      {
        stdio: 'pipe',
        cwd: tempDir,
        maxBuffer: 20 * 1024 * 1024
      }
    );
  } catch (e) {
    const stderr = e && e.stderr ? e.stderr.toString() : '';
    const stdout = e && e.stdout ? e.stdout.toString() : '';
    throw new Error(`Failed to build C# harness: ${stderr || stdout || e.message}`);
  }

  const dll = path.join(tempDir, 'out', 'Harness.dll');

  return {
    run(mode, body) {
      try {
        const out = execFileSync(dotnet, [dll, mode], {
          input: JSON.stringify(body == null ? null : body),
          encoding: 'utf8',
          cwd: tempDir,
          maxBuffer: 20 * 1024 * 1024
        });
        return { ok: true, out };
      } catch (e) {
        return {
          ok: false,
          err:
            (e && e.stderr ? e.stderr.toString() : '') ||
            (e && e.stdout ? e.stdout.toString() : '') ||
            e.message
        };
      }
    }
  };
}

let runner = null;
try {
  runner = compileHarness();
  logWarn('C# harness compiled successfully');
} catch (e) {
  logWarn(`C# harness unavailable: ${e.message}`);
}

app.get('/health', (req, res) => {
  res.type('text/plain').status(200).send('ok');
});

function textReply(res, status, body) {
  res.type('text/plain').status(status).send(String(body));
}

function route(mode) {
  return (req, res) => {
    try {
      if (!runner) {
        return textReply(res, 500, 'harness not available');
      }

      const result = runner.run(mode, req.body);
      if (!result.ok) {
        return textReply(res, 500, result.err);
      }

      return textReply(res, 200, result.out);
    } catch (e) {
      return textReply(res, 500, e && e.message ? e.message : String(e));
    }
  };
}

app.post('/harness/createorderdraftcommandhandler-handle', route('createorderdraftcommandhandler-handle'));
app.post('/harness/basketitemextensions-toorderitemdto', route('basketitemextensions-toorderitemdto'));
app.post('/harness/basketitemextensions-toorderitemsdto', route('basketitemextensions-toorderitemsdto'));

const port = parseInt(process.env.PORT || '3001', 10);
app.listen(port, '0.0.0.0', () => {
  console.log(`Harness listening on ${port}`);
});

File: harness.cs
using System.Reflection;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using eShop.EventBus.Abstractions;
using eShop.EventBus.Events;
using eShop.IntegrationEventLogEF.Services;
using eShop.Ordering.API.Application.Commands;
using eShop.Ordering.API.Application.IntegrationEvents;
using eShop.Ordering.API.Application.Models;
using eShop.Ordering.API.Application.Queries;
using eShop.Ordering.API.Infrastructure.Services;
using eShop.Ordering.Domain.AggregatesModel.BuyerAggregate;
using eShop.Ordering.Domain.AggregatesModel.OrderAggregate;
using eShop.Ordering.Domain.Seedwork;
using eShop.Ordering.Infrastructure;
using eShop.Ordering.Infrastructure.Repositories;
using MediatR;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using CardTypeQueryModel = eShop.Ordering.API.Application.Queries.CardType;
using QueryOrder = eShop.Ordering.API.Application.Queries.Order;

var builder = WebApplication.CreateBuilder(args);

builder.WebHost.UseUrls($"http://0.0.0.0:{Environment.GetEnvironmentVariable("PORT") ?? "3001"}");

builder.Services.AddRouting();
builder.Services.AddLogging();
builder.Services.ConfigureHttpJsonOptions(options =>
{
    options.SerializerOptions.PropertyNameCaseInsensitive = true;
    options.SerializerOptions.DefaultIgnoreCondition = JsonIgnoreCondition.Never;
});

var connectionString =
    Environment.GetEnvironmentVariable("ConnectionStrings__OrderingDB") ??
    "Host=orderingdb;Database=OrderingDB;Username=postgres;Password=yourWeak(!)Password";

builder.Services.AddDbContext<OrderingContext>(options =>
{
    options.UseNpgsql(connectionString);
});

var app = builder.Build();

app.MapGet("/health", () => Results.Text("ok", "text/plain"));

static IResult Plain(object? value, int statusCode = 200)
{
    string text;
    if (value is null)
    {
        text = "null";
    }
    else if (value is string s)
    {
        text = s;
    }
    else
    {
        text = JsonSerializer.Serialize(value);
    }

    return Results.Text(text, "text/plain", Encoding.UTF8, statusCode);
}

static async Task<T?> ReadJsonBodyAsync<T>(HttpRequest request)
{
    return await JsonSerializer.DeserializeAsync<T>(
        request.Body,
        new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true
        });
}

static string ExceptionText(Exception ex)
{
    var baseEx = ex.GetBaseException();
    return $"{baseEx.GetType().FullName}: {baseEx.Message}";
}

static T SetProperty<T>(T instance, string propertyName, object? value)
{
    var prop = typeof(T).GetProperty(propertyName, BindingFlags.Public | BindingFlags.Instance);
    prop?.SetValue(instance, value);
    return instance;
}

static CreateOrderCommand BuildCreateOrderCommandFromSample(CreateOrderCommandRequest dto)
{
    var command = new CreateOrderCommand();

    SetProperty(command, nameof(CreateOrderCommand.UserId), dto.UserId);
    SetProperty(command, nameof(CreateOrderCommand.UserName), dto.UserName);
    SetProperty(command, nameof(CreateOrderCommand.City), dto.City);
    SetProperty(command, nameof(CreateOrderCommand.Street), dto.Street);
    SetProperty(command, nameof(CreateOrderCommand.State), dto.State);
    SetProperty(command, nameof(CreateOrderCommand.Country), dto.Country);
    SetProperty(command, nameof(CreateOrderCommand.ZipCode), dto.ZipCode);
    SetProperty(command, nameof(CreateOrderCommand.CardNumber), dto.CardNumber);
    SetProperty(command, nameof(CreateOrderCommand.CardHolderName), dto.CardHolderName);
    SetProperty(command, nameof(CreateOrderCommand.CardExpiration), dto.CardExpiration);
    SetProperty(command, nameof(CreateOrderCommand.CardSecurityNumber), dto.CardSecurityNumber);
    SetProperty(command, nameof(CreateOrderCommand.CardTypeId), dto.CardTypeId);

    var field = typeof(CreateOrderCommand).GetField("_orderItems", BindingFlags.NonPublic | BindingFlags.Instance);
    field?.SetValue(command, dto.OrderItems ?? new List<OrderItemDTO>());

    return command;
}

static async Task SeedIfNeededAsync(OrderingContext db)
{
    await db.Database.EnsureCreatedAsync();

    if (!await db.CardTypes.AnyAsync())
    {
        db.CardTypes.AddRange(
            new CardType { Id = 1, Name = "Visa" },
            new CardType { Id = 2, Name = "MasterCard" },
            new CardType { Id = 3, Name = "Amex" }
        );
    }

    if (!await db.Buyers.AnyAsync(b => b.IdentityGuid == "dast-user"))
    {
        db.Buyers.Add(new Buyer("dast-user", "DAST User"));
    }

    if (!await db.Orders.AnyAsync(o => o.Id == 123))
    {
        var address = new Address("1 Main St", "Testville", "CA", "US", "12345");
        var order = new eShop.Ordering.Domain.AggregatesModel.OrderAggregate.Order(
            "dast-user",
            "alice",
            address,
            1,
            "XXXXXXXXXXXX1111",
            "123",
            "Alice",
            DateTime.UtcNow.AddYears(2)
        );
        order.AddOrderItem(10, "Widget", 12.5m, 0m, "https://example.test/img.png", 2);
        db.Orders.Add(order);
    }

    await db.SaveChangesAsync();
}

var loaded = new Dictionary<string, bool>();

void SafeMap(string key, Action mapper)
{
    try
    {
        mapper();
        loaded[key] = true;
        app.Logger.LogInformation("Loaded harness route for {Key}", key);
    }
    catch (Exception ex)
    {
        loaded[key] = false;
        app.Logger.LogWarning(ex, "Failed to load harness route for {Key}", key);
    }
}

SafeMap("CreateOrderDraftCommandHandler.Handle", () =>
{
    app.MapPost("/harness/createorderdraftcommandhandler-handle", async (HttpRequest request) =>
    {
        try
        {
            var payload = await ReadJsonBodyAsync<CreateOrderDraftHandleRequest>(request) ?? new CreateOrderDraftHandleRequest();
            var handler = new CreateOrderDraftCommandHandler();
            var result = await handler.Handle(
                payload.Message ?? new CreateOrderDraftCommand("user-123", new List<BasketItem>()),
                CancellationToken.None);
            return Plain(result);
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

SafeMap("BasketItemExtensions.ToOrderItemDTO", () =>
{
    app.MapPost("/harness/basketitemextensions-toorderitemdto", async (HttpRequest request) =>
    {
        try
        {
            var item = await ReadJsonBodyAsync<BasketItem>(request) ?? new BasketItem();
            var result = eShop.Ordering.API.Extensions.BasketItemExtensions.ToOrderItemDTO(item);
            return Plain(result);
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

SafeMap("BasketItemExtensions.ToOrderItemsDTO", () =>
{
    app.MapPost("/harness/basketitemextensions-toorderitemsdto", async (HttpRequest request) =>
    {
        try
        {
            var items = await ReadJsonBodyAsync<List<BasketItem>>(request) ?? new List<BasketItem>();
            var result = eShop.Ordering.API.Extensions.BasketItemExtensions.ToOrderItemsDTO(items).ToList();
            return Plain(result);
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

SafeMap("OrderQueries.Handle", () =>
{
    app.MapGet("/harness/orderqueries-handle", async (HttpRequest request, OrderingContext db) =>
    {
        try
        {
            await SeedIfNeededAsync(db);
            var idText = request.Query["id"].FirstOrDefault() ?? "123";
            var id = int.Parse(idText);
            var queries = new OrderQueries(db);
            QueryOrder result = await queries.GetOrderAsync(id);
            return Plain(result);
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

SafeMap("OrderQueries.GetOrdersFromUserAsync", () =>
{
    app.MapGet("/harness/orderqueries-getordersfromuserasync", async (HttpRequest request, OrderingContext db) =>
    {
        try
        {
            await SeedIfNeededAsync(db);
            var userId = request.Query["userId"].FirstOrDefault() ?? "dast-user";
            var queries = new OrderQueries(db);
            var result = await queries.GetOrdersFromUserAsync(userId);
            return Plain(result);
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

SafeMap("OrderQueries.GetCardTypesAsync", () =>
{
    app.MapGet("/harness/orderqueries-getcardtypesasync", async (OrderingContext db) =>
    {
        try
        {
            await SeedIfNeededAsync(db);
            var queries = new OrderQueries(db);
            IEnumerable<CardTypeQueryModel> result = await queries.GetCardTypesAsync();
            return Plain(result);
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

SafeMap("CreateOrderCommandHandler.Handle", () =>
{
    app.MapPost("/harness/createordercommandhandler-handle", async (HttpRequest request, OrderingContext db, ILoggerFactory loggerFactory) =>
    {
        try
        {
            await SeedIfNeededAsync(db);

            var payload = await ReadJsonBodyAsync<CreateOrderCommandHandleRequest>(request) ?? new CreateOrderCommandHandleRequest();
            var repository = new OrderRepository(db);
            var mediator = new NoopMediator();
            var identityService = new StubIdentityService();
            var eventService = new StubOrderingIntegrationEventService();
            var logger = loggerFactory.CreateLogger<CreateOrderCommandHandler>();

            var handler = new CreateOrderCommandHandler(mediator, eventService, repository, identityService, logger);
            var command = BuildCreateOrderCommandFromSample(payload.Message ?? new CreateOrderCommandRequest());
            var result = await handler.Handle(command, CancellationToken.None);
            return Plain(new
            {
                Result = result,
                SavedEvents = eventService.SavedEvents.Select(e => new { e.Id, Type = e.GetType().FullName }).ToList()
            });
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

SafeMap("OrderingIntegrationEventService.AddAndSaveEventAsync", () =>
{
    app.MapPost("/harness/orderingintegrationeventservice-addandsaveeventasync", async (HttpRequest request, OrderingContext db, ILoggerFactory loggerFactory) =>
    {
        try
        {
            await db.Database.EnsureCreatedAsync();

            var evt = await ReadJsonBodyAsync<IntegrationEvent>(request) ?? new IntegrationEvent();
            var logger = loggerFactory.CreateLogger<OrderingIntegrationEventService>();
            var eventBus = new StubEventBus();
            var eventLogService = new StubIntegrationEventLogService();
            var service = new OrderingIntegrationEventService(eventBus, db, eventLogService, logger);

            using var tx = await db.Database.BeginTransactionAsync();
            typeof(OrderingContext)
                .GetField("_currentTransaction", BindingFlags.NonPublic | BindingFlags.Instance)
                ?.SetValue(db, tx);

            await service.AddAndSaveEventAsync(evt);

            return Plain(new
            {
                EventId = evt.Id,
                SavedCount = eventLogService.Saved.Count
            });
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

SafeMap("CancelOrderCommandHandler.Handle", () =>
{
    app.MapPut("/harness/cancelordercommandhandler-handle", async (HttpRequest request, OrderingContext db) =>
    {
        try
        {
            await SeedIfNeededAsync(db);

            var payload = await ReadJsonBodyAsync<CancelOrderHandleRequest>(request) ?? new CancelOrderHandleRequest();
            var repository = new OrderRepository(db);
            var handler = new CancelOrderCommandHandler(repository);
            var result = await handler.Handle(payload.Command ?? new CancelOrderCommand(123), CancellationToken.None);
            return Plain(result);
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

SafeMap("ShipOrderCommandHandler.Handle", () =>
{
    app.MapPut("/harness/shipordercommandhandler-handle", async (HttpRequest request, OrderingContext db) =>
    {
        try
        {
            await SeedIfNeededAsync(db);

            var payload = await ReadJsonBodyAsync<ShipOrderHandleRequest>(request) ?? new ShipOrderHandleRequest();
            var order123 = await db.Orders.FindAsync((payload.Command?.OrderNumber) ?? 123);
            if (order123 != null)
            {
                try
                {
                    order123.SetAwaitingValidationStatus();
                    order123.SetStockConfirmedStatus();
                    order123.SetPaidStatus();
                    await db.SaveChangesAsync();
                }
                catch
                {
                }
            }

            var repository = new OrderRepository(db);
            var handler = new ShipOrderCommandHandler(repository);
            var result = await handler.Handle(payload.Command ?? new ShipOrderCommand(123), CancellationToken.None);
            return Plain(result);
        }
        catch (Exception ex)
        {
            return Plain(ExceptionText(ex), 500);
        }
    });
});

app.Run();

public sealed class CreateOrderDraftHandleRequest
{
    public CreateOrderDraftCommand? Message { get; set; }
    public string? CancellationToken { get; set; }
}

public sealed class CreateOrderCommandHandleRequest
{
    public CreateOrderCommandRequest? Message { get; set; }
    public string? CancellationToken { get; set; }
}

public sealed class CancelOrderHandleRequest
{
    public CancelOrderCommand? Command { get; set; }
    public string? CancellationToken { get; set; }
}

public sealed class ShipOrderHandleRequest
{
    public ShipOrderCommand? Command { get; set; }
    public string? CancellationToken { get; set; }
}

public sealed class CreateOrderCommandRequest
{
    public string UserId { get; set; } = "user-123";
    public string UserName { get; set; } = "alice";
    public string City { get; set; } = "Testville";
    public string Street { get; set; } = "1 Main St";
    public string State { get; set; } = "CA";
    public string Country { get; set; } = "US";
    public string ZipCode { get; set; } = "12345";
    public string CardNumber { get; set; } = "XXXXXXXXXXXX1111";
    public string CardHolderName { get; set; } = "Alice";
    public DateTime CardExpiration { get; set; } = DateTime.Parse("2030-01-01T00:00:00Z").ToUniversalTime();
    public string CardSecurityNumber { get; set; } = "123";
    public int CardTypeId { get; set; } = 1;
    public List<OrderItemDTO> OrderItems { get; set; } = new()
    {
        new OrderItemDTO
        {
            ProductId = 10,
            ProductName = "Widget",
            UnitPrice = 12.5m,
            Discount = 0m,
            Units = 2,
            PictureUrl = "https://example.test/img.png"
        }
    };
}

public sealed class StubIdentityService : IIdentityService
{
    public string GetUserIdentity() => "dast-user";
    public string GetUserName() => "DAST User";
}

public sealed class StubOrderingIntegrationEventService : IOrderingIntegrationEventService
{
    public List<IntegrationEvent> SavedEvents { get; } = new();

    public Task AddAndSaveEventAsync(IntegrationEvent evt)
    {
        SavedEvents.Add(evt);
        return Task.CompletedTask;
    }

    public Task PublishEventsThroughEventBusAsync(Guid transactionId) => Task.CompletedTask;
}

public sealed class StubIntegrationEventLogService : IIntegrationEventLogService
{
    public List<(IntegrationEvent Event, object? Transaction)> Saved { get; } = new();

    public Task<IEnumerable<eShop.IntegrationEventLogEF.IntegrationEventLogEntry>> RetrieveEventLogsPendingToPublishAsync(Guid transactionId)
        => Task.FromResult<IEnumerable<eShop.IntegrationEventLogEF.IntegrationEventLogEntry>>(Array.Empty<eShop.IntegrationEventLogEF.IntegrationEventLogEntry>());

    public Task SaveEventAsync(IntegrationEvent @event, Microsoft.EntityFrameworkCore.Storage.IDbContextTransaction transaction)
    {
        Saved.Add((@event, transaction));
        return Task.CompletedTask;
    }

    public Task MarkEventAsPublishedAsync(Guid eventId) => Task.CompletedTask;
    public Task MarkEventAsInProgressAsync(Guid eventId) => Task.CompletedTask;
    public Task MarkEventAsFailedAsync(Guid eventId) => Task.CompletedTask;
}

public sealed class StubEventBus : IEventBus
{
    public Task PublishAsync(IntegrationEvent @event) => Task.CompletedTask;
}

public sealed class NoopMediator : IMediator
{
    public Task Publish(object notification, CancellationToken cancellationToken = default) => Task.CompletedTask;
    public Task Publish<TNotification>(TNotification notification, CancellationToken cancellationToken = default)
        where TNotification : INotification => Task.CompletedTask;
    public Task<TResponse> Send<TResponse>(IRequest<TResponse> request, CancellationToken cancellationToken = default)
        => Task.FromResult(default(TResponse)!);
    public Task Send<TRequest>(TRequest request, CancellationToken cancellationToken = default)
        where TRequest : IRequest => Task.CompletedTask;
    public IAsyncEnumerable<object?> CreateStream(object request, CancellationToken cancellationToken = default)
        => AsyncEnumerable.Empty<object?>();
    public IAsyncEnumerable<TResponse> CreateStream<TResponse>(IStreamRequest<TResponse> request, CancellationToken cancellationToken = default)
        => AsyncEnumerable.Empty<TResponse>();
}
